### PR TITLE
new modules for managing dell me4 storage arrays

### PR DESCRIPTION
**Pull request by acalm**
##### SUMMARY
New modules for managing Dell ME4 storage arrays, as there's no generic Dell community collection i figured these modules would fit under general storage
##### ISSUE TYPE

- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME
* dell_me4_disk_group
* dell_me4_initiator
* dell_me4_map_volume
* dell_me4_volume
* dell_me4_info
* dell_me4_iscsi_config

##### ADDITIONAL INFORMATION
Provided modules have been tested with a ME4024 array. Since the arrays are proprietary there's no way to provide any useful tests other than simple linting of the modules

### CODE REVIEW

File: changelogs/fragments/dell_me4_modules.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+minor_changes:
+  - new storage modules for Dell ME4 arrays
```
Comment (felixfontein) [inline changelogs/fragments/dell_me4_modules.yml:3]: New modules should not have a changelog fragment.

File: plugins/modules/storage/dell_emc/dell_me4_disk_group.py [hunk 2]
```diff
@@ -0,0 +1,501 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright (C) 2020, Andreas Calminder <andreas.calminder@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+ANSIBLE_METADATA = {
```
Comment (felixfontein) [inline plugins/modules/storage/dell_emc/dell_me4_disk_group.py:10]: Please remove this.

File: plugins/modules/storage/dell_emc/dell_me4_disk_group.py [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+    'metadata_version': '0.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = '''
+---
+module: dell_me4_disk_group
+short_description: Manage disk groups in a Dell EMC me4 series SAN
```
Comment (felixfontein) [inline plugins/modules/storage/dell_emc/dell_me4_disk_group.py:19]:
```suggestion
short_description: Manage disk groups in a Dell EMC me4 series SAN
version_added: 1.3.0
```

**Review by felixfontein — COMMENTED** [review]
Please note that all new modules must have tests.

Also, I think it's better to create a new collection for these modules instead of adding them here. CC @gundalow

**Comment by acalm** [comment]
@felixfontein :
> Please note that all new modules must have tests.
Agreed, however I cannot provide any integration tests as the arrays are built on proprietary software, I'll provide some unit tests

> Also, I think it's better to create a new collection for these modules instead of adding them here. CC @gundalow
Yes, you're probably right. My reasoning for putting them here was that I'm in no way affiliated with the vendor, other than writing the modules, creating a vendor connected collection didn't seem right. I'll create a new collection if you think it's the best way, although I'd appreciate some help with the naming as I don't want to "steal" any vendor namespaces or make it seem like some officially supported tools

**Review by Andersson007 — COMMENTED** [review]
@acalm thanks for the modules

I checked one module by script and it showed that there are several issues related to documentation standards:
```
description: line 1 does not end with a dot
opt username: line 1 does not start with a capital letter
opt username: line 1 does not end with a dot
opt spare_disks: line 1 does not start with a capital letter
opt spare_disks: line 1 does not end with a dot
opt disks: line 1 does not start with a capital letter
opt disks: line 1 does not end with a dot
opt disks: line 2 does not start with a capital letter
opt disks: line 2 does not end with a dot
opt disks: line 3 does not start with a capital letter
opt disks: line 3 does not end with a dot
opt disks: line 4 does not start with a capital letter
opt disks: line 4 does not end with a dot
opt disks: line 5 does not start with a capital letter
opt disks: line 5 does not end with a dot
opt disks: line 6 does not start with a capital letter
opt disks: line 6 does not end with a dot
opt raid: line 1 does not start with a capital letter
opt raid: line 1 does not end with a dot
opt raid: line 2 does not start with a capital letter
opt raid: line 2 does not end with a dot
opt name: line 1 does not start with a capital letter
opt name: line 1 does not end with a dot
opt name: line 2 does not start with a capital letter
opt name: line 2 does not end with a dot
opt adapt_spare_capacity: line 1 does not start with a capital letter
opt adapt_spare_capacity: line 1 does not end with a dot
opt adapt_spare_capacity: line 2 does not start with a capital letter
opt adapt_spare_capacity: line 2 does not end with a dot
opt adapt_spare_capacity: line 3 does not start with a capital letter
opt adapt_spare_capacity: line 3 does not end with a dot
opt adapt_spare_capacity: line 4 does not end with a dot
opt hostname: line 1 does not start with a capital letter
opt hostname: line 1 does not end with a dot
opt controller: line 1 does not start with a capital letter
opt state: line 1 does not start with a capital letter
opt state: line 1 does not end with a dot
opt state: line 2 does not start with a capital letter
opt state: line 2 does not end with a dot
opt state: line 3 does not start with a capital letter
opt state: line 3 does not end with a dot
opt verify_cert: line 1 does not start with a capital letter
opt verify_cert: line 1 does not end with a dot
opt mode: line 1 does not start with a capital letter
opt mode: line 1 does not end with a dot
opt mode: line 0 contains 208 characters which seems to be too long
opt chunk_size: line 1 does not start with a capital letter
opt chunk_size: line 1 does not end with a dot
opt chunk_size: line 2 does not start with a capital letter
opt chunk_size: line 2 does not end with a dot
opt storage_type: line 1 does not start with a capital letter
opt storage_type: line 1 does not end with a dot
opt password: line 1 does not start with a capital letter
opt password: line 1 does not end with a dot
opt pool: line 1 does not start with a capital letter
opt pool: line 1 does not end with a dot
notes: line 1 does not end with a dot
notes: check_mode support is not mentioned
examples: "- name" of #1 does not start with a capital letter
examples: "- name" of #2 does not start with a capital letter
examples: "- name" of #3 does not start with a capital letter
return: no RETURN section, there must be, at least, RETURN = r"""#"""
```
Please, scrutinize [ansible doc **standards]()** and make **all** the modules matching them

**Comment by Andersson007** [comment]
Also please create unit tests for all the modules where possible

**Comment by Andersson007** [comment]
> Please note that all new modules must have tests.
> 
> Also, I think it's better to create a new collection for these modules instead of adding them here. CC @gundalow

Definitely!

**Comment by acalm** [comment]
Thanks @Andersson007, for some reason I couldn't see those errors while
testing locally, I just got the message about output in stderr being
considered an error. I'll see to them and create unit tests as well. As
mentioned by others before, I'll try and make this a collection of it's own

**Comment by Andersson007** [comment]
@acalm those things aren't covered by tests but they mentioned in the documentations (see the provided reference).
Maybe later we'll add the checks to sanity tests, will see.
Thanks!

**Comment by Andersson007** [comment]
@acalm ah, sorry, i noticed that i didn't past it there, fixed. https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html

**Comment by acalm** [comment]
@Andersson007 thanks for the help! I've created separate [collection](https://github.com/acalm/unknown.storage) for the modules, with fixes for the above issues. This PR can be closed

**Comment by felixfontein** [comment]
@acalm you might find the collection template useful: https://github.com/ansible-collections/collection_template/

It contains some useful things that are easy to re-use, including a workflow for GitHub Actions (https://github.com/ansible-collections/collection_template/blob/main/.github/workflows/ansible-test.yml) for sanity tests, unit tests, and integration tests (modular, i.e. you can only take the ones your collection needs).

**Comment by Andersson007** [comment]
@acalm cool, thanks for the information:) @felixfontein thanks for reviewing!
closed via the above

**Pull request closed without merging**